test/cpp/jit/test_gpu_validator.h

```diff
@@ -0,0 +1,321 @@
+
+  double base_half_abs_tol = -1;
+  double base_half_rel_tol = -1;
+  double base_float_abs_tol = -1;
+  double base_float_rel_tol = -1;
+};
+
+namespace {
+
+// Returns abs and relative values to use for validation
+std::pair<double, double> getTolerance(
+    DataType dtype,
+    int64_t reduction_size,
+    const ValidationConstants& tolerances) {
+  switch (dtype) {
+    case DataType::Float: {
+      const auto& sum_tolerance_entry = tolerances.sum_tolerances_float;
+      const auto& base_abs = tolerances.base_float_abs_tol;
+      const auto& base_rel = tolerances.base_float_rel_tol;
+
+      if (reduction_size <= 1) {
+        // No reduction case
+        if (base_abs == -1 || base_rel == -1) {
+          return {sum_tolerance_entry[0][1], sum_tolerance_entry[1][1]};
+        } else {
+          return {base_abs, base_rel};
+        }
+      } else {
+        // Reduction case
+        size_t entry = 0;
+        while (sum_tolerance_entry[entry][0] < reduction_size &&
+               entry < sum_tolerance_entry.size()) {
+          entry++;
+        }
+        double abs_tol = 0.0;
+        if (entry + 1 < sum_tolerance_entry.size()) {
+          // Grab the next entry up so we have some margin
+          abs_tol = sum_tolerance_entry[entry + 1][1];
+        } else {
+          // If we hit the end of the list, return twice the max error we
+          // measured
+          abs_tol = sum_tolerance_entry[sum_tolerance_entry.size() - 1][1] * 2.;
+        }
+        // Relative tol we're going to set to 1% of abs tol just for
+        // a small margin of rel error.
+        return {abs_tol, abs_tol * 0.01};
+      }
+    }
+    case DataType::Half: {
+      // Copied from float case
+      const auto& sum_tolerance_entry = tolerances.sum_tolerances_half;
+      const auto& base_abs = tolerances.base_half_abs_tol;
+      const auto& base_rel = tolerances.base_half_rel_tol;
+
+      if (reduction_size <= 1) {
+        // No reduction case
+        if (base_abs == -1 || base_rel == -1) {
+          return {sum_tolerance_entry[0][1], sum_tolerance_entry[1][1]};
+        } else {
+          return {base_abs, base_rel};
+        }
+      } else {
+        // Reduction case
+        size_t entry = 0;
+        while (sum_tolerance_entry[entry][0] < reduction_size &&
+               entry < sum_tolerance_entry.size()) {
+          entry++;
+        }
+        double abs_tol = 0.0;
+        if (entry + 1 < sum_tolerance_entry.size()) {
+          // Grab the next entry up so we have some margin
+          abs_tol = sum_tolerance_entry[entry + 1][1];
+        } else {
+          // If we hit the end of the list, return twice the max error we
+          // measured
+          abs_tol = sum_tolerance_entry[sum_tolerance_entry.size() - 1][1] * 2.;
+        }
+        // Relative tol we're going to set to 1% of abs tol just for
+        // a small margin of rel error.
+        return {abs_tol, abs_tol * 0.01};
```

Comment:
from PyTorch: [TORCH.ALLCLOSE](https://pytorch.org/docs/stable/generated/torch.allclose.html)
```
∣input−other∣ ≤ atol+rtol×∣other∣
```
I don't believe `rtol` is meant to scale with `atol` .

Also I use to think having more reduction elements shouldn't require relaxing `rtol`.

But could be I'm not understanding this scheme.

Reply:
I was going back and forth on this, was thinking to have rtol as 0 and just use atol, but that seemed strange to me, so just put a small amount of rtol.